### PR TITLE
fix: support estimator output path parameterization 

### PR DESCRIPTION
**Pull request by jerrypeng7773**
*Issue #, if available:*

#3104
#3078

*Description of changes:*

support estimator `output_path` parameterization 

*Testing done:*

unit + integ

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by qidewenwhen — APPROVED** [review]

File: src/sagemaker/estimator.py [hunk 1]
```diff
@@ -695,14 +695,19 @@ def _stage_user_code_in_s3(self) -> str:
 
         Returns: S3 URI
         """
-        local_mode = self.output_path.startswith("file://")
+        local_mode = not is_pipeline_variable(self.output_path) and self.output_path.startswith(
```
Comment (navaj0) [inline src/sagemaker/estimator.py:698]: I guess this works, but the statement is confusing. Whether it is local mode shouldn't depends on whether the output path is a pipeline variable or not. 

The follow up if elif block is really hard to read. 

How about just do this
```
if is_pipeline_variable(self.output_path):
    if self.code_location is None:
         ...
    else:
         ...
else: 
```

File: src/sagemaker/estimator.py [hunk 2]
```diff
@@ -695,26 +695,41 @@ def _stage_user_code_in_s3(self) -> str:
 
         Returns: S3 URI
         """
-        local_mode = self.output_path.startswith("file://")
-
-        if self.code_location is None and local_mode:
+        if is_pipeline_variable(self.output_path):
             code_bucket = self.sagemaker_session.default_bucket()
```
Comment (qidewenwhen) [inline src/sagemaker/estimator.py:699]: Nit: we can move it into the line 700 `if` block

File: src/sagemaker/estimator.py [hunk 3]
```diff
@@ -695,26 +695,41 @@ def _stage_user_code_in_s3(self) -> str:
...
+                    self.output_kms_key
+                    if code_bucket == self.sagemaker_session.default_bucket()
+                    else None
+                )
         else:
-            code_bucket, key_prefix = parse_s3_url(self.code_location)
-            code_s3_prefix = "/".join(filter(None, [key_prefix, self._current_job_name, "source"]))
+            local_mode = self.output_path.startswith("file://")
+            if self.code_location is None:
+                code_bucket, _ = parse_s3_url(self.output_path)
```
Comment (qidewenwhen) [inline src/sagemaker/estimator.py:716]: Seems we're changing the original logics.
```
if self.code_location is None and local_mode:
    code_bucket = self.sagemaker_session.default_bucket()
    ....
```

File: src/sagemaker/estimator.py [hunk 4]
```diff
@@ -695,26 +695,41 @@ def _stage_user_code_in_s3(self) -> str:
...
-            output_bucket, _ = parse_s3_url(self.output_path)
-            kms_key = self.output_kms_key if code_bucket == output_bucket else None
+                output_bucket, _ = parse_s3_url(self.output_path)
+                kms_key = (
+                    None
+                    if local_mode
+                    else self.output_kms_key
+                    if code_bucket == output_bucket
+                    else None
+                )
```
Comment (qidewenwhen) [inline src/sagemaker/estimator.py:732]: Nit:  it's a bit hard to follow. Can we add a () so it looks like:
```
kms_key = (
                    None
                    if local_mode
                    else (self.output_kms_key
                    if code_bucket == output_bucket
                    else None)
                )
```
Or expend this to 
```
if local_mode:
 kms_key = ...
elif code_bucket == output_bucket:
 kms_key = ...
else:
 kms_key = ...
```

File: src/sagemaker/fw_utils.py [hunk 5]
```diff
@@ -395,8 +397,10 @@ def model_code_key_prefix(code_location_key_prefix, model_name, image):
     Returns:
         str: the key prefix to be used in uploading code
     """
-    training_job_name = sagemaker.utils.name_from_image(image)
-    return "/".join(filter(None, [code_location_key_prefix, model_name or training_job_name]))
+    name_from_image = f"/model_code/{int(time.time())}"
+    if not is_pipeline_variable(image):
+        name_from_image = sagemaker.utils.name_from_image(image)
+    return "/".join(filter(None, [code_location_key_prefix, model_name or name_from_image]))
```
Comment (qidewenwhen) [inline src/sagemaker/fw_utils.py:403]: Seems this should belong to another PR: https://github.com/aws/sagemaker-python-sdk/pull/3112 right?

File: src/sagemaker/estimator.py (hunk 4 above)
Comment (jerrypeng7773) [reply src/sagemaker/estimator.py]: this is failing local_mode, got a new revision.

File: src/sagemaker/fw_utils.py (hunk 5 above)
Comment (jerrypeng7773) [reply src/sagemaker/fw_utils.py:403]: sigh ... yea, master branch gets messed up.

**Review by nmadan — APPROVED** [review]

**Review by navinsoni — APPROVED** [review]

**Review by shreyapandit — APPROVED** [review]
